drgn/helpers/linux/module.py

```diff
@@ -0,0 +1,228 @@
+        # layout and any memory which could be freed after init. The init_layout
+        # is usually NULL / size 0. The module_layout structure has more
+        # information to say where text ends, where rodata ends, etc. We ignore
+        # these.
+        core = _range_from_module_layout(mod.core_layout)
+        init = _range_from_module_layout(mod.init_layout)
+    except AttributeError:
+        # Prior to 4.5, 7523e4dc5057 ("module: use a structure to encapsulate
+        # layout."), the layout information was stored as variables directly in
+        # the struct module. They were prefixed with "core_" and "init_".
+        core = _range_from_module(mod, "core")
+        init = _range_from_module(mod, "init")
+
+    ret = [core]
+    if init:
+        ret.append(init)
+    return ret
+
+
+def _addrmod_tree(mod_tree: Object, addr: int) -> Object:
+    prog = mod_tree.prog_
+
+    # The module tree is "latched": there are two parallel trees. Which one is
+    # in use depends on the seqcount, which gets incremented for each
+    # modification. This is a really neat approach that allows reads in parallel
+    # with a writer. In our use case, it's probably not worth verifying the
+    # seqcount after the fact. What we do need is the index (0 or 1). This may
+    # be a seqcount_latch_t, or before 24bf401cebfd6 ("rbtree_latch: Use
+    # seqcount_latch_t"), a regular seqcount_t.
+    try:
+        idx = int(mod_tree.root.seq.seqcount.sequence & 1)
+    except AttributeError:
+        idx = int(mod_tree.root.seq.sequence & 1)
+
+    # In ac3b432839234 ("module: replace module_layout with module_memory"),
+    # struct module_layout was replaced by module_memory. The module_layout
+    # encoded the separate regions (text, data, rodata, etc) in a single
+    # structure, whereas module_memory is a simple base pointer followed by a
+    # size: one module_memory structure is used per kind of memory. However,
+    # both of them contain a "base" pointer that indicates the start of the
+    # region, a "size" that indicates its total size, and a "mtn.mod" pointer
+    # which refers to the relevant module. So for our use case, they are
+    # interchangeable, except for their names.
+    try:
+        tp = prog.type("struct module_memory")
+    except LookupError:
+        tp = prog.type("struct module_layout")
+
+    def cmp(v: int, node: Object) -> int:
+        start = node.base.value_()
+        end = start + node.size.value_()
+        if v < start:
+            return -1
+        elif v >= end:
+            return 1
+        else:
+            return 0
+
+    mem = rb_find(
+        tp,
+        mod_tree.root.tree[idx].address_of_(),
+        f"mtn.node.node[{idx}]",  # container_of allows array indices!
+        addr,
+        cmp,
+    )
+    if mem:
+        return mem.mtn.mod
+    else:
+        return NULL(prog, "struct module *")
+
+
+@takes_program_or_default
+def address_to_module(prog: Program, addr: IntegerLike) -> Object:
```

Comment:
The equivalent kernel function is [`__module_address()`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/module.h?h=v6.10-rc7#n614), but that name doesn't make much sense to me, so this is fine.